File: bound/golang/crypto/jwk.go
package crypto

import (
	"web5/web5c"
)

type JWK struct {
	ALG string `json:"alg,omitempty"`
	KTY string `json:"kty,omitempty"`
	CRV string `json:"crv,omitempty"`
	D   string `json:"d,omitempty"`
	X   string `json:"x,omitempty"`
	Y   string `json:"y,omitempty"`
}

func (j JWK) ComputeThumbprint() (string, error) {
	cJWK := j.ToCJWK()
	thumbprint := cJWK.ComputeThumbprint()
	return thumbprint, nil
}

func (j JWK) ToCJWK() *web5c.CJWK {
	return &web5c.CJWK{
		ALG: j.ALG,
		KTY: j.KTY,
		CRV: j.CRV,
		D:   j.D,
		X:   j.X,
		Y:   j.Y,
	}
}

func NewJWKFromCJWK(cJWK *web5c.CJWK) *JWK {
	return &JWK{
		ALG: cJWK.ALG,
		KTY: cJWK.KTY,
		CRV: cJWK.CRV,
		D:   cJWK.D,
		X:   cJWK.X,
		Y:   cJWK.Y,
	}
}

File: bound/typescript/src/crypto/jwk.ts
import { catchWeb5Error } from "../errors";
import wasm from "../wasm"

export class Jwk {
  public alg: string | undefined;
  public kty: string;
  public crv: string;
  public d: string | undefined;
  public x: string;
  public y: string | undefined;

  static fromWasmJwk(wasmJwk: wasm.WasmJwk): Jwk {
    return new Jwk(
      wasmJwk.alg,
      wasmJwk.kty,
      wasmJwk.crv,
      wasmJwk.d,
      wasmJwk.x,
      wasmJwk.y
    );
  }

  constructor(
    alg: string | undefined,
    kty: string,
    crv: string,
    d: string | undefined,
    x: string,
    y: string | undefined
  ) {
    this.alg = alg;
    this.kty = kty;
    this.crv = crv;
    this.d = d;
    this.x = x;
    this.y = y;
  }

  toWasmJwk(): wasm.WasmJwk {
    return new wasm.WasmJwk(this.alg, this.kty, this.crv, this.d, this.x, this.y);
  }

  computeThumbprint(): string {
    try {
      const wasmJwk = this.toWasmJwk();
      return wasmJwk.compute_thumbprint();
    } catch (error) {
      throw catchWeb5Error(error);
    }
  }
}